infrastructures/docker/src/main/java/org/eclipse/che/workspace/infrastructure/docker/DockerInternalRuntime.java

```diff
@@ -203,29 +208,39 @@ protected void internalStart(Map<String, String> startOptions) throws Infrastruc
             }
         } catch (InfrastructureException | InterruptedException | RuntimeException e) {
             boolean interrupted = e instanceof InterruptedException || Thread.interrupted();
-            boolean runtimeDestroyingNeeded = !startSynchronizer.isStopCalled();
-            if (runtimeDestroyingNeeded) {
-                try {
-                    destroyRuntime(null);
-                } catch (Exception destExc) {
-                    LOG.error(destExc.getLocalizedMessage(), destExc);
-                }
+            try {
+                destroyRuntime(null);
+            } catch (Exception destExc) {
+                LOG.error(destExc.getLocalizedMessage(), destExc);
             }
             if (interrupted) {
-                throw new InfrastructureException("Docker runtime start was interrupted");
+                InfrastructureException ex = new RuntimeStartInterruptedException(getContext().getIdentity());
+                startSynchronizer.exception = ex;
+                throw ex;
             }
+            startSynchronizer.exception = e;
             if (e instanceof InfrastructureException) {
                 throw (InfrastructureException)e;
-            } else {
-                throw new InternalInfrastructureException(e.getLocalizedMessage(), e);
             }
+            throw new InternalInfrastructureException(e.getLocalizedMessage(), e);
+        } finally {
+            startSynchronizer.release();
         }
     }
 
     @Override
     protected void internalStop(Map<String, String> stopOptions) throws InfrastructureException {
-        startSynchronizer.interruptStartThread();
-        destroyRuntime(stopOptions);
+        if (startSynchronizer.startThread != null) {
```

Comment:
`startThread` field neither locked nor volatile, can this check fail because of caching of pointer value?